### ISSUE TITLE
Window 's show and hide event

### ISSUE THREAD

**Issue by stepjacky** [nwjs/nw.js#4731]
All version 
windows 7,8 ,10 
when create a window ,usually use show and hide method to display or not
i need do somthing on a window show ,or hide ,but can't found any event responsed to the two method 
can some else instead!?

**Comment by 1j01** [comment]
The best you can currently do is probably to overwrite and wrap the `show` and `hide` methods, e.g.
```js
var _show = win.show;
var _hide = win.hide;
win.show = function(){
    win.apply(_show);
    handleWindowShown();
};
win.hide = function(){
    win.apply(_hide);
    handleWindowHidden();
};
```

**Comment by stepjacky** [comment]
ofcause to do like this ,but i feeled  indecently,overwriting the original method is not a good way!

**Comment by 1j01** [comment]
@stepjacky Right, there really should be events, and maybe an `isHidden` [method](https://github.com/nwjs/nw.js/issues/4809).

**Comment by fritx** [comment]
In 0.12.3, I used to use `win.on('focus')` & `win.on('blur')` to modify a state `win._isFocus`, it was fine until I migrated to  0.18.8.

The old handler would be overriden by a new one, which is weird and should be fixed.

**Comment by fritx** [comment]
I ended up with making the following patch upon `win.on/removeListener`:

```js
// 自实现并重写 win.on/removeListener
// 修复nw focus/blur事件侦听 会被重写
// https://github.com/nwjs/nw.js/issues/4731#issuecomment-262722326
win._handlers = []
win._on = win.on
win.on = (e, fn) => {
  if (win._handlers[e]) {
    win._handlers[e].push(fn)
  } else {
    win._handlers[e] = [fn]
  }
  win.remove
  win.removeAllListeners(e) // 进一步确保清空
  win._on(e, (...args) => {
    const arr = win._handlers[e]
    if (arr) {
      arr.forEach(f => {
        f(...args)
      })
    }
  })
}
win._removeListener = win.removeListener
win.removeListener = (e, fn) => {
  const arr = win._handlers[e]
  if (arr) {
    const i = arr.indexOf(fn)
    if (i > -1) arr.splice(i, 1)
  }
}
```

**Comment by EthraZa** [comment]
Today I found out that the HTML5 notificaton popup will not show over a fullscreen app. So I guess that a good solution might be to quickly focus my nwjs app call the notification and blur it away. The problem now is that by the lack of an isHidden or other similar method, I can't know if my nwjs app is minimized, maximized or hidden, to make it go back to the previous state after call the notification.

So, I would like to ask for properties of the window that tell us what is the real state of it. 
We already have win.isFullscreen, now we need win.isHidden and win.isMinimized or something like that.
